src/libraries/System.Private.CoreLib/src/System/Half.cs

```diff
@@ -0,0 +1,633 @@
+        }
+
+        public static Half Parse(string s, NumberStyles style)
+        {
+            return (Half)float.Parse(s, style);
+        }
+
+        public static Half Parse(string s, IFormatProvider provider)
+        {
+            return (Half)float.Parse(s, provider);
+        }
+
+        public static Half Parse(string s, NumberStyles style = DefaultParseStyle, IFormatProvider? provider = null)
+        {
+            if (s is null)
+            {
+                throw new ArgumentNullException(nameof(s));
+            }
+            return Parse(s.AsSpan(), style, provider);
+        }
+
+        public static Half Parse(ReadOnlySpan<char> s, NumberStyles style = DefaultParseStyle, IFormatProvider? provider = null)
+        {
+            return (Half)(float.Parse(s, style, provider));
+        }
+
+        public static bool TryParse(string s, out Half result)
+        {
+            return TryParse(s, DefaultParseStyle, provider: null, out result);
+        }
+
+        public static bool TryParse(ReadOnlySpan<char> s, out Half result)
+        {
+            return TryParse(s, DefaultParseStyle, provider: null, out result);
+        }
+
+        public static bool TryParse(string s, NumberStyles style, IFormatProvider? provider, out Half result)
+        {
+            return TryParse(s.AsSpan(), style, provider, out result);
+        }
+
+        public static bool TryParse(ReadOnlySpan<char> s, NumberStyles style, IFormatProvider? provider, out Half result)
+        {
+            bool ret = false;
+            if (float.TryParse(s, style, provider, out float floatResult))
+            {
+                result = (Half)floatResult;
+                ret = true;
+            }
+            else
+            {
+                result = default(Half);
+            }
+            return ret;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool AreZero(Half left, Half right)
+        {
+            // IEEE defines that positive and negative zero are equal, this gives us a quick equality check
+            // for two values by or'ing the private bits together and stripping the sign. They are both zero,
+            // and therefore equivalent, if the resulting value is still zero.
+            return (ushort)((left.m_value | right.m_value) & ~SignMask) == 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsNaNOrZero(Half value)
+        {
+            return ((value.m_value - 1) & ~SignMask) >= PositiveInfinityBits;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ushort StripSign(Half value)
+        {
+            return (ushort)(value.m_value & ~SignMask);
+        }
+
+        public int CompareTo(object? obj)
+        {
+            if (!(obj is Half))
+            {
+                return (obj is null) ? 1 : throw new ArgumentException(SR.Arg_MustBeHalf);
```

Comment:
I noticed that some of our other structs could move the null check inside the type check as you have done here.
Eg., Int32.cs
```c#
        public int CompareTo(object? value)
        {
            if (value == null)
            {
                return 1;
            }

            // NOTE: Cannot use return (_value - value) as this causes a wrap
            // around in cases where _value - value > MaxValue.
            if (value is int i)
            {
```